numpy/tests/test_public_api.py

```diff
@@ -547,3 +548,88 @@ def test_core_shims_coherence():
 
         else:
             assert member is getattr(core, member_name)
+
+
+def test_functions_single_location():
+    """
+    Check that each public function is available from one location only.
+
+    Test performs BFS search traversing NumPy's public API. It flags
+    any function-like object that is accessible from more that one place.
+    """
+    from typing import Any, Callable, Dict, List, Set, Tuple
+    from numpy._core._multiarray_umath import (
+        _ArrayFunctionDispatcher as dispatched_function
+    )
+
+    visited_modules: Set[types.ModuleType] = {np}
+    visited_functions: Set[Callable[..., Any]] = set()
+    # Functions often have `__name__` overridden, therefore we need
+    # to keep track of locations where functions have been found.
+    functions_original_paths: Dict[Callable[..., Any], str] = dict()
+
+    # Here we aggregate functions with more than one location.
+    # It must be empty for the test to pass.
+    duplicated_functions: List[Tuple] = []
+
+    modules_queue = [np]
+
+    while len(modules_queue) > 0:
+
+        module = modules_queue.pop()
+
+        for member_name in dir(module):
+            member = getattr(module, member_name)
+
+            # first check if we got a module
+            if (
+                inspect.ismodule(member) and  # it's a module
+                "numpy" in member.__name__ and  # inside NumPy
+                not member_name.startswith("_") and  # not private
+                "numpy._core" not in member.__name__ and  # outside _core
+                # not a legacy or testing module
+                member_name not in ["f2py", "ma", "testing", "tests"] and
+                # skip numpy.polynomial
```

Comment:
Maybe expand this comment to explain why we're skipping `polynomial`? I don't remember offhand so future readers definitely won't

Reply:
I revisited `numpy.polynomial` and I decided to include it in the test. 

There's one function, `numpy.polynomial.polyutils.trimcoef` that is duplicated in other polynomial submodules on purpose. (as `lagtrim` in `numpy.polynomial.laguerre`, `hermtrim` in `numpy.polynomial.hermite`, etc.)

I manually skip that one function instead.